server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -218,44 +224,91 @@ public void onFailure(Exception e) {
                         logger.warn("Recovery failed", e);
                         // we reset `recovered` in the listener don't reset it here otherwise there might be a race
                         // that resets it to false while a new recover is already running?
-                        recoveryListener.onFailure("state recovery failed: " + e.getMessage());
+                        GatewayService.this.onFailure("state recovery failed: " + e.getMessage());
                     }
 
                     @Override
-                    protected void doRun() throws Exception {
-                        gateway.performStateRecovery(recoveryListener);
+                    protected void doRun() {
+                        recoveryRunnable.run();
                     }
                 });
             }
         }
     }
 
+    private void onFailure(String message) {
+        recovered.set(false);
+        scheduledRecovery.set(false);
+        // don't remove the block here, we don't want to allow anything in such a case
+        logger.info("metadata state not restored, reason: {}", message);
+    }
+
+    class RecoverStateUpdateTask extends ClusterStateUpdateTask {
+
+        @Override
+        public ClusterState execute(ClusterState currentState) {
+            return allocationService.reroute(updateRoutingTable(removeStateNotRecoveredBlock(currentState)), "state recovered");
+        }
+
+        private ClusterState updateRoutingTable(ClusterState state) {
+            // initialize all index routing tables as empty
+            RoutingTable.Builder routingTableBuilder = RoutingTable.builder(state.routingTable());
+            for (ObjectCursor<IndexMetaData> cursor : state.metaData().indices().values()) {
+                routingTableBuilder.addAsRecovery(cursor.value);
+            }
+            // start with 0 based versions for routing table
+            routingTableBuilder.version(0);
+            return ClusterState.builder(state).routingTable(routingTableBuilder.build()).build();
+        }
+
+        private ClusterState removeStateNotRecoveredBlock(ClusterState state) {
+            return ClusterState.builder(state)
+                    .blocks(ClusterBlocks.builder()
+                            .blocks(state.blocks()).removeGlobalBlock(STATE_NOT_RECOVERED_BLOCK).build())
+                    .build();
+        }
+
+        @Override
+        public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+            logger.info("routing table cluster state update is successfully processed");
```

Comment:
Previous log message was "recovered [{}] indices into cluster_state", which I think is suitable when Zen1 discovery is used (and I override `clusterStateProcessed` method for Zen1 and log old error message), but less suitable for Zen2. Because in Zen2 we already have `ClusterState` in place and just want to update the routing table.

Reply:
Actually, I've restored previous message even for Zen2 case during rework